src/axom/klee/GeometryOperators.hpp

```diff
@@ -0,0 +1,390 @@
+// Copyright (c) 2017-2021, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#ifndef AXOM_KLEE_GEOMETRYOPERATOR_HPP
+#define AXOM_KLEE_GEOMETRYOPERATOR_HPP
+
+#include <memory>
+#include <vector>
+
+#include "axom/core/numerics/Matrix.hpp"
+#include "axom/klee/Dimensions.hpp"
+#include "axom/klee/Geometry.hpp"
+#include "axom/klee/Units.hpp"
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/geometry/Vector.hpp"
+
+namespace axom
+{
+namespace klee
+{
+class GeometryOperatorVisitor;
+
+/**
+ * A GeometryOperator describes an operation to perform on the Geometry
+ * of a Shape.
+ *
+ * There is a subclass of GeometryOperator for each operator defined in
+ * the format specification. You can figure out which one you're working
+ * with via the accept() method. This follows the standard visitor pattern.
+ */
+class GeometryOperator
+{
+public:
+  /**
+   * Create an operator with the given start properties
+   * \param startProperties the properties before the operator is applied
+   */
+  explicit GeometryOperator(const TransformableGeometryProperties &startProperties);
+
+  virtual ~GeometryOperator() = default;
+
+  /**
+   * Get the properties that the operator expects to start in
+   *
+   * \return the properties which must be true before this operator is applied
+   */
+  const TransformableGeometryProperties &getStartProperties() const
+  {
+    return m_startProperties;
+  }
+
+  /**
+   * Get the properties after this operator is applied
+   *
+   * \return the properties which are true after this operator is applied
+   */
+  virtual TransformableGeometryProperties getEndProperties() const
+  {
+    // Be default, end properties are the same as start properties, so
+    // this is correct.
+    return m_startProperties;
+  }
+
+  /**
+   * Accept the given visitor. The appropriate visit() method will
+   * be called on the visitor based on the run-time type of this operator.
+   *
+   * \param visitor the visitor to accept.
+   */
+  virtual void accept(GeometryOperatorVisitor &visitor) const = 0;
+
+private:
+  TransformableGeometryProperties m_startProperties;
+};
+
+/**
+ * A MatrixOperator is a type of GeometryOperator whose operation can be
+ * expressed as a 4x4 affine transformation matrix.
+ */
+class MatrixOperator : public GeometryOperator
+{
+public:
+  using GeometryOperator::GeometryOperator;
+
+  /**
+   * Convert this operator to its matrix representation.
+   *
+   * \return a 4x4 affine transformation matrix
+   */
+  virtual numerics::Matrix<double> toMatrix() const = 0;
+};
+
+/**
+ * A CompositeOperator is a GeometryOperator which consists of a list of
+ * other operators.
+ */
+class CompositeOperator : public GeometryOperator
+{
+public:
+  using GeometryOperator::GeometryOperator;
+
+  using OpPtr = std::shared_ptr<const GeometryOperator>;
```

Comment:
Likewise here: I think we could have `CompositeOperator` "own" the operators it is given, so `OpPtr` could be made `unique_ptr`.

Reply:
See above. Ownership can actually be shared with named operators.